File: 01_algorithm/sort/06_merge/merge-sort.py
def binarySearch(list, target):
    left = 0
    right = len(list) - 1

    while left <= right:
        mid = left + (right - left) // 2
        midVal = list[mid]
        if midVal == target:
            return mid
        elif midVal < target:
            left = mid + 1
        else:
            right = mid - 1

    # 基底ケース
    return -1


print(binarySearch([1], 1))  # 0
print(binarySearch([1, 2, 3, 4, 5, 6, 7, 8, 9, 10], 3))  # 2
print(binarySearch([1, 2, 3, 4, 5, 6, 7, 8, 9, 10], 9))  # 8
print(binarySearch([1, 3, 5, 7, 9], 5))  # 2
print(binarySearch([2, 4, 6, 8, 10], 2))  # 0
print(binarySearch([1, 2, 3, 4, 5], 5))  # 4
print(binarySearch([10, 20, 30, 40, 50], 25))  # -1
print(binarySearch([], 1))  # -1
print(binarySearch([4], 4))  # 0
print(binarySearch([4], 3))  # -1

File: 01_algorithm/sort/06_merge/merge_sort.dart
import '../../../00_util/dart/generate.dart';

void mergeSort(List<int> nums){
  var l = nums.length;
  if (l <= 1){
    return;
  }

  var mid = l ~/2;
  // var list = List<Type>.filled(size, initialValue);
  var(left,right) = (nums.sublist(0, mid),nums.sublist(mid, l));

  mergeSort(left);
  mergeSort(right);

  merge(nums,left,right);
}

void merge(List<int> nums,List<int> left, List<int> right){
  var (i,j,k) = (0,0,0);
  while (i < left.length && j < right.length){
    var(lv,rv) = (left[i],right[j]);
    if (lv < rv){
      nums[k] = lv;
      i++;
    }else{
      nums[k] = rv;
      j++;
    }
    k++;
  }
  while (i < left.length){
    nums[k] = left[i];
    i++;
    k++;
  }
  while (j < right.length){
    nums[k] = right[j];
    j++;
    k++;
  }
}

void main(){
  var nums1 = GenerateRandomArray(10);
  mergeSort(nums1);
  print(nums1);
}